edit.ml

```diff
@@ -110,61 +110,16 @@ let split_edge instrs pc label pc' =
   new_instrs, pc_map
 
 let replace_uses_in_instruction old_name new_name instr : instruction =
-  let in_simple_expression (exp:simple_expression) : simple_expression =
-    match exp with
-    | Constant _ -> exp
-    | Var x -> if x = old_name then Var new_name else exp in
-  let in_expression exp : expression =
-    match exp with
-    | Simple se -> Simple (in_simple_expression se)
-    | Op (op, exps) ->
-      Op (op, List.map in_simple_expression exps) in
-  let in_arg e : argument = in_expression e in
-  let in_osr osr : osr_def =
-    match osr with
-    | Osr_var (x, exp) -> Osr_var (x, in_expression exp)
-  in
-  match instr with
-  | Call (x, f, exs) ->
-    assert(x <> old_name);   (* -> invalid scope *)
-    Call (x, in_expression f, List.map in_arg exs)
-  | Stop e ->
-    Stop (in_expression e)
-  | Return e ->
-    Return (in_expression e)
-  | Decl_var (x, exp) ->
-    assert(x <> old_name);   (* -> invalid scope *)
-    Decl_var (x, in_expression exp)
-  | Decl_array (x, def) ->
-    assert (x <> old_name);
-    let def = match def with
-      | Length e -> Length (in_expression e)
-      | List es -> List (List.map in_expression es)
-    in Decl_array (x, def)
-  | Assign (x, exp) ->
-    Assign (x, in_expression exp)
-  | Array_assign (x, index, exp) ->
-    let x' = if x = old_name then new_name else x in
-    Array_assign (x', in_expression index, in_expression exp)
-  | Drop x ->
-    if x = old_name then Drop new_name else instr
-  | Print exp ->
-    Print (in_expression exp)
-  | Assert exp ->
-    Assert (in_expression exp)
-  | Branch (exp, l1, l2) ->
-    Branch (in_expression exp, l1, l2)
-  | Osr {cond; target; map} ->
-    let cond = List.map in_expression cond in
-    let map = List.map in_osr map in
-    Osr {cond; target; map}
-  | Read x ->
-    assert (x <> old_name);
-    instr
-
-  | Label _ | Goto _ | Comment _ ->
-    assert (VarSet.is_empty (used_vars instr));
-    instr
+  let mapper = object (self)
+    inherit Instr.mapper as super
+    method! bound_variable x =
+      if x = old_name then new_name else x
+    method! binder x = x
+    method! instruction = function[@warning "-4"]
+      | Assign (x, e) -> Assign (x, self#expression e)
```

Comment:
Note: this special-case behavior of *not* touching the assigned variable was kept from the previous version of the code, but I don't know if it's still necessary. (There is a call after this mapper is used that works on Assign instructions specifically, and it may have to be changed in synch.). It would be nicer (more regular) to get rid of it.